contracts/governance/StakingRewards/StakingRewards.sol

```diff
@@ -0,0 +1,152 @@
+		stopBlock = block.number;
+	}
+
+	/**
+	 * @notice Sets the max duration
+	 * @dev Rewards can be collected for a maximum duration at a time. This
+	 * is to avoid Block Gas Limit failures. Setting it zero would mean that it will loop
+	 * through the entire duration since the start of rewards program.
+	 * It should ideally be set to a value, for which the rewards can be easily processed.
+	 * @param _duration Max duration for which rewards can be collected at a go (in seconds)
+	 * */
+	function setMaxDuration(uint256 _duration) public onlyOwner {
+		maxDuration = _duration;
+	}
+
+	/**
+	 * @notice Collect rewards
+	 * @dev User calls this function to collect SOV staking rewards as per the SIP-0024 program.
+	 * The weighted stake is calculated using getPriorWeightedStake. Block number sent to the functon
+	 * must be a finalised block, hence we deduct 1 from the current block. User is only allowed to withdraw
+	 * after intervals of 14 days.
+	 * */
+	function collectReward() external {
+		uint256 count;
+		uint256 weightedStake;
+		uint256 lastFinalisedBlock = block.number - 1;
+		uint256 currentTS = block.timestamp;
+		address staker = msg.sender;
+		if (withdrawals[staker] == 0) {
+			require((currentTS.sub(startTime)) > TWO_WEEKS, "allowed after 14 days of start");
+			withdrawals[staker] = startTime;
+		} else {
+			require(currentTS > withdrawals[staker], "allowed after 14 days");
+		}
+
+		for (uint256 i = withdrawals[staker]; i < currentTS && i < withdrawals[staker].add(maxDuration); i += TWO_WEEKS) {
+			count++;
+			weightedStake = weightedStake.add(_computeRewardForDate(staker, lastFinalisedBlock, i));
+		}
+		require(weightedStake > 0, "weightedStake is zero");
+		withdrawals[staker] += count.mul(TWO_WEEKS);
+		_payReward(staker, weightedStake);
+	}
+
+	/**
+	 * @notice Internal function to calculate weighted stake
+	 * @dev If the rewards program is stopped, the user will still continue to
+	 * earn till the end of staking period based on the stop block.
+	 * @param _staker Staker address
+	 * @param _block Last finalised block
+	 * @param _date The date to compute prior weighted stakes
+	 * */
+	function _computeRewardForDate(
+		address _staker,
+		uint256 _block,
+		uint256 _date
+	) internal view returns (uint256 weightedStake) {
+		weightedStake = staking.getPriorWeightedStake(_staker, _block, _date);
+		if (stopBlock > 0) {
+			uint256 previousWeightedStake = staking.getPriorWeightedStake(_staker, stopBlock, _date);
+			if (previousWeightedStake < weightedStake) {
+				weightedStake = previousWeightedStake;
+			}
+		}
+	}
+
+	/**
+	 * @notice Internal function to pay rewards
+	 * @dev Base rate is annual, but we pay interest for 14 days,
+	 * which is 1/26 of one staking year (1092 days)
+	 * @param _staker User address
+	 * @param weightedStake the weighted stake
+	 * */
+	function _payReward(address _staker, uint256 weightedStake) internal {
+		uint256 amount = weightedStake.mul(BASE_RATE).div(DIVISOR);
```

Comment:
there is no unit tests verifying that the amounts paid are actually correct. please verify the amounts are the same as calculated here: https://docs.google.com/spreadsheets/d/1BfghMoIen2CIUTu-f0gD_kcPhwgI7tTdRFue-VpVGUM/edit#gid=0